Use angular-gettext to translate interface strings

asaunier opened the pull request:
Use built-in angular-gettext [1] to translate the interface strings in french.

Some points still need to be fixed:
* [x] form input placeholders or logo title
* [ ] ng-messages (form validation errors)
* [x] values of ng-repeat combos (for ex in editing form)
* [x] strings containing Mako variables (for ex ``${total} waypoints``) 
* [x] use doc culture as ``<html lang="{lang}">`` instead of interface culture
* [x] add interface culture selector

Related to https://github.com/c2corg/v6_ui/issues/6

[1] https://angular-gettext.rocketeer.be/dev-guide/getting-started/

asaunier commented:
This PR is ready for reviewing.

It uses the angular-gettext based translation system that comes with pyramid_closure. Pages are translated on the fly using javascript replacements and lang detection. The server side is not aware of the user culture at all. Which is consistent with the choice of having a stateless server side UI, knowing nothing about the user (including his ID). The pages will be generated only once for all languages (and most likely cached using Varnish).

Drawbacks are that the client browser has more work to do to render the page (really a problem?) and that JSON translation files (downloaded by angular-gettext when the page lang is changed) might grow a little big depending on the numbers of strings to translate.

See my instance for a demo: http://c2corgv6-demo.gis.internal:6555/

I have addressed most of the TODO items even though I have not succeeded in having the form validation error messages translated yet (?) (pb with ng-messages?).

The "lang" attribute in ``<html lang="...">`` is now set to the document culture currently shown, for "view" pages. For other pages (list, editing => not reachable by search engine anyway), "lang" is set to "mainCtrl.lang". I just realize that I have dropped this property from the main controller but anyway I am not sure it is really relevent to have such a value for "lang" since Angular properties are not supposed to be interpreted by search engines => should I remove "lang" for those pages?

In the server generated HTML, interface strings are then in english. Is it a problem for SEO if the interface uses a different language then the content? We could minimize the problem (and reduce a bit the pages sizes and i18n dictionaries sizes) by using minimalistic msgids instead of real sentences/expressions (EN actually has a translation file).

I have moved most of the lang-related code from App.MainController to App.LangController. A "translate" function remains for the "ng-options" translations (or maybe I can use the "translate" function from langCtrl?). App.MainController is now a little empty. I don't know yet what we will use it for, maybe for linking the different subcontrollers together?

Points that still need to be covered:
* currently the user browser languages prefs are used to figure the best culture, for every single page => we should do it only once and then save the browsing language to some web storage about the user (probably something to link with the user auth).
* the culture dropdown value is not correctly initialized even though the user prefs are actually detected. Don't know why the dropdown is not updated using its ng-model?!

gberaudo commented on c2corg_ui/static/js/maincontroller.js:
```diff
@@ -5,45 +5,25 @@ goog.require('app');
-  this.gettextCatalog_.loadRemote(
-      this.langUrlTemplate_.replace('__lang__', lang));
-  this.lang = lang;
+  $rootScope['translate'] = function(str) {
```
In the `langController` you are prefixing the translate function by the controller name.
Would it work to put it on the main controller and prefix by the main controller name instead?

asaunier replied:
The function is indeed duplicated. That would indeed make sense to have it only once.
I have tried to use ``$rootScope.translate()`` in the langDirective but with no luck. Your solution might work, actually I would like to find a way to completely remove the culture-related code from the main controller and have every thing in the langController. Maybe I could simply move ``$rootScope['translate'] = function(str) {`` in the langController instead of the main controller?

gberaudo commented on c2corg_ui/templates/base.html:
```diff
@@ -25,10 +25,10 @@
     <ul>
-      <li><a href="${request.route_url('waypoints_index')}">Waypoints</a></li>
-      <li><a href="${request.route_url('routes_index')}">Routes</a></li>
+      <li><a href="${request.route_url('waypoints_index')}" translate>Waypoints</a></li>
```
It is better to use `ng-href`.

asaunier replied:
Why? This is a standard link, not an expression. And it's better to use HTML links so that search engines may access those pages.

gberaudo replied:
Right, thought it was an angular expression.

tsauerwein commented:
This is looking good, thanks!

asaunier commented:
In the last commits I have removed ``lang="{mainCtrl.lang}"`` from the html opening tag (still set to the doc culture for the main doc pages) since mainCtrl has no longer a "lang" property and that this value does not seem that relevent anyway.

I have removed the "translate()" function from the MainController, only using the one available in LangController. MainController is now an empty shell that does nothing... I guess we might need it later so I don't remove it yet.

Merging when Travis approves.

asaunier commented:
> I have removed the "translate()" function from the MainController, only using the one available in LangController.

Oops, actually it doesn't work. I had to add back a translate() function to the MainController:
https://github.com/c2corg/v6_ui/commit/92a985cdaa28eda3ab0f99e8d637ec7adf18213e

I still can't figure out why the lang dropdown is not updated when the user's browser lang prefs are detected. The ``<select>``tag actually has a ``ng-model="langCtrl.culture"``attribute and ``this.culture is updated when the user's lang is detected: 
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/lang.js#L64

Is there something missing somewhere? A $scope.$apply() for instance?
